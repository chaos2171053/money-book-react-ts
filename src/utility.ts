export const LIST_VIEW = 'list'
export const CHART_VIEW = 'chart'
export const TYPE_INCOME = 'income'
export const TYPE_OUTCOME = 'outcome'

export const padLeft = (n: number) => {
    return n < 10 ? '0' + n : n
}

export const range = (size: number, startAt: number = 0) => {
    const arr = []
    for (let i = 0; i < size; i++) {
        arr[i] = startAt + i
    }
    return arr
}

export const parseToYearAndMonth = (str?: string) => {
    const date = str ? new Date(str) : new Date()
    return {
        year: date.getFullYear(),
        month: date.getMonth() + 1,
    }
}

export const isValidDate = (dateString: string) => {
    const regEx = /^\d{4}-\d{2}-\d{2}$/;
    if (!dateString.match(regEx)) return false;  // Invalid format
    const d = new Date(dateString);
    if (Number.isNaN(d.getTime())) return false; // Invalid date
    return d.toISOString().slice(0, 10) === dateString;
}