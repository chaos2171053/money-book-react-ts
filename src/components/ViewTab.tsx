import React from 'react'
import { LIST_VIEW, CHART_VIEW } from '../utility'
import Ionicon from 'react-ionicons'

interface IProps {
    activeTab: string,
    onTabChange: (view: string) => void
}

const generateLinkClass = (current: string, view: string) => {
    return (current === view) ? 'nav-link active' : 'nav-lick'
}

const ViewTab = (props: IProps) => (
    <ul className='nav nav-tabs nav-fill my-4'>
        <li className='nav-item'>
            <a
                className={generateLinkClass(props.activeTab, LIST_VIEW)}
                href='javascript;'
                onClick={(event) => { event.preventDefault(); props.onTabChange(LIST_VIEW) }}
            >
                <Ionicon
                    className="rounded-circle mr-2"
                    fontSize="25px"
                    color={'#007bff'}
                    icon='ios-paper'
                />
                列表模式
                </a>
        </li>
        <li className='nav-item'>
            <a
                className={generateLinkClass(props.activeTab, CHART_VIEW)}
                href='javascript;'
                onClick={(event) => { event.preventDefault(); props.onTabChange(CHART_VIEW) }}>
                <Ionicon
                    className="rounded-circle mr-2"
                    fontSize="25px"
                    color={'#007bff'}
                    icon='ios-pie'
                />
                图表模式
                </a>
        </li>
    </ul>
)

export default ViewTab