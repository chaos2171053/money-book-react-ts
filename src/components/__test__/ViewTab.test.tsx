import React from 'react'
import { shallow } from 'enzyme'
import ViewTab from '../ViewTab'
import { LIST_VIEW } from '../../utility'


const props = {
    activeTab: LIST_VIEW,
    onTabChange: jest.fn()
}

let wrapper: any
describe('test ViewTab component', () => {
    beforeEach(() => {
        wrapper = shallow(<ViewTab {...props} />)
    })
    it('should render the component to match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
    it('should render correct tab items length ', () => {
        const navItems = wrapper.find('.nav-item')
        expect(navItems.length).toEqual(2)
    })
    // it('should trigger the correct function callbacks', () => {

    //     const navItems = wrapper.find('.nav-item')

    //     expect(navItems.first().find('a').first().hasClass('active')).toEqual(true)
    //     expect(navItems.last().find('a').first().hasClass('active')).toEqual(false)

    //     wrapper.find('.active').first().simulate('click', { preventDefault: () => { } })
    //     //navItems.last().find('a').first().simulate('click', { preventDefault: () => { } })

    //     expect(navItems.first().find('a').first().hasClass('active')).toEqual(false)
    //     expect(navItems.last().find('a').first().hasClass('active')).toEqual(true)

    // })

})