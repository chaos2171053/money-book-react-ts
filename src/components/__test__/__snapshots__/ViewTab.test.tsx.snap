// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ViewTab component should render the component to match snapshot 1`] = `ShallowWrapper {}`;
